File: ports/python/hqtui/widgets/controls.py
"""Buttons, inputs, tabs and the overlays that sit above a whole screen."""

from __future__ import annotations

from dataclasses import dataclass, replace
from enum import Enum
from typing import Sequence

from ..buffer import Attrs, Style
from ..color import Color
from ..surface import BorderStyle, BoxOptions, Surface, TextOptions
from ..theme import elevate
from ..unicode import Align, fit, string_width, truncate, wrap

__all__ = [
    "ButtonOptions",
    "ButtonVariant",
    "CheckboxOptions",
    "CheckboxVariant",
    "CommandPaletteOptions",
    "ModalButton",
    "ModalOptions",
    "PaletteItem",
    "SelectOptions",
    "TabVariant",
    "TabsOptions",
    "TextInputOptions",
    "TooltipOptions",
    "draw_button",
    "draw_checkbox",
    "draw_command_palette",
    "draw_modal",
    "draw_select",
    "draw_tabs",
    "draw_text_input",
    "draw_tooltip",
]


class ButtonVariant(str, Enum):
    PRIMARY = "primary"
    SUCCESS = "success"
    WARNING = "warning"
    DANGER = "danger"
    GHOST = "ghost"


@dataclass(frozen=True, slots=True)
class ButtonOptions:
    label: str = ""
    focused: bool = False
    color: Color | None = None
    variant: "ButtonVariant | str" = ButtonVariant.PRIMARY
    disabled: bool = False
    width: int | None = None
    align: "Align | str" = Align.CENTER


def _variant_color(surface: Surface, options: ButtonOptions) -> Color:
    if options.color is not None:
        return options.color
    t = surface.theme
    return {
        ButtonVariant.SUCCESS: t.success,
        ButtonVariant.WARNING: t.warning,
        ButtonVariant.DANGER: t.danger,
        ButtonVariant.GHOST: t.muted,
    }.get(options.variant, t.primary)


def draw_button(surface: Surface, options: ButtonOptions) -> int:
    """Returns the width drawn."""
    if surface.empty:
        return 0
    theme = surface.theme
    color = _variant_color(surface, options)
    label = f" {options.label} "
    width = min(
        options.width if options.width is not None else string_width(label), surface.width
    )

    if options.disabled:
        style = Style(fg=theme.muted, bg=elevate(theme, 0.05))
    elif options.focused:
        style = Style(
            fg=theme.background if theme.dark else theme.surface, bg=color, attrs=Attrs.BOLD
        )
    elif options.variant == ButtonVariant.GHOST:
        style = Style(fg=color)
    else:
        style = Style(fg=color, bg=theme.surface.mix(color, 0.16), attrs=Attrs.BOLD)

    surface.text(
        0, 0, fit(truncate(label, width), width, options.align),
        TextOptions(fg=style.fg, bg=style.bg, attrs=style.attrs),
    )
    return width


class CheckboxVariant(str, Enum):
    """Render as a box, a switch, or a radio dot."""

    CHECKBOX = "checkbox"
    TOGGLE = "toggle"
    RADIO = "radio"


@dataclass(frozen=True, slots=True)
class CheckboxOptions:
    label: str = ""
    checked: bool = False
    focused: bool = False
    color: Color | None = None
    variant: "CheckboxVariant | str" = CheckboxVariant.CHECKBOX


def draw_checkbox(surface: Surface, options: CheckboxOptions) -> int:
    """Returns the width drawn."""
    if surface.empty:
        return 0
    theme = surface.theme
    color = options.color
    if color is None:
        color = theme.success if options.checked else theme.muted

    if options.variant == CheckboxVariant.TOGGLE:
        glyph = "[▮ ]" if options.checked else "[ ▮]"
    elif options.variant == CheckboxVariant.RADIO:
        glyph = "(●)" if options.checked else "( )"
    else:
        glyph = "[✓]" if options.checked else "[ ]"

    attrs = Attrs.BOLD if options.focused else 0
    x = surface.text(0, 0, glyph, TextOptions(fg=color, attrs=attrs))
    if options.label:
        x += surface.text(
            x, 0, f" {options.label}",
            TextOptions(
                fg=theme.foreground if options.focused else theme.muted, attrs=attrs
            ),
        )
    return x


@dataclass(frozen=True, slots=True)
class SelectOptions:
    value: str = ""
    focused: bool = False
    open: bool = False
    options: Sequence[str] = ()
    selected_index: int = 0
    width: int | None = None
    color: Color | None = None


def draw_select(surface: Surface, options: SelectOptions) -> None:
    """A closed dropdown, or an open one with its option list underneath."""
    if surface.empty:
        return
    theme = surface.theme
    width = min(options.width if options.width is not None else surface.width, surface.width)
    color = options.color
    if color is None:
        color = theme.border_focused if options.focused else theme.border
    field_bg = elevate(theme, 0.05)
    label = " " + truncate(options.value, max(0, width - 4))
    surface.text(
        0, 0, fit(label, max(0, width - 2), Align.LEFT),
        TextOptions(
            fg=theme.foreground, bg=field_bg, attrs=Attrs.BOLD if options.focused else 0
        ),
    )
    surface.text(
        width - 2, 0, " ▴" if options.open else " ▾", TextOptions(fg=color, bg=field_bg)
    )

    if options.open and options.options:
        height = min(len(options.options), max(0, surface.height - 1))
        list_bg = elevate(theme, 0.08)
        for i in range(height):
            selected = i == options.selected_index
            surface.text(
                0, i + 1,
                fit(" " + truncate(options.options[i], max(0, width - 2)), width, Align.LEFT),
                TextOptions(
                    fg=theme.selection_text if selected else theme.foreground,
                    bg=theme.selection if selected else list_bg,
                ),
            )


@dataclass(frozen=True, slots=True)
class TextInputOptions:
    value: str = ""
    placeholder: str = ""
    focused: bool = False
    cursor: int | None = None
    """Caret index; None means the end of the value."""
    width: int | None = None
    label: str = ""
    password: bool = False
    color: Color | None = None


def draw_text_input(surface: Surface, options: TextInputOptions) -> None:
    if surface.empty:
        return
    theme = surface.theme
    width = min(options.width if options.width is not None else surface.width, surface.width)
    label_width = string_width(options.label) + 1 if options.label else 0
    if options.label:
        surface.text(0, 0, options.label, TextOptions(fg=theme.muted))
    field_width = max(0, width - label_width)
    bg = elevate(theme, 0.1 if options.focused else 0.05)
    surface.fill_rect(label_width, 0, field_width, 1, Style(bg=bg))

    # The reference counts UTF-16 units for the password mask and the default
    # caret; counting codepoints is the same for every value a person types and
    # is what a Python caller would expect.
    shown = "•" * len(options.value) if options.password else options.value
    empty = shown == ""
    text = options.placeholder if empty else shown
    surface.text(
        label_width + 1, 0, truncate(text, max(0, field_width - 2)),
        TextOptions(fg=theme.muted if empty else theme.foreground, bg=bg),
    )

    if options.focused:
        caret = options.cursor if options.cursor is not None else string_width(shown)
        cursor_x = min(label_width + 1 + caret, label_width + max(0, field_width - 1))
        surface.style_rect(
            cursor_x, 0, 1, 1,
            Style(
                fg=theme.background,
                bg=options.color if options.color is not None else theme.cursor,
            ),
        )


class TabVariant(str, Enum):
    FILLED = "filled"
    UNDERLINE = "underline"
    """Underline the active tab instead of filling it."""


@dataclass(frozen=True, slots=True)
class TabsOptions:
    tabs: Sequence[str] = ()
    active: int = 0
    color: Color | None = None
    align: "Align | str" = Align.LEFT
    variant: "TabVariant | str" = TabVariant.FILLED


def draw_tabs(surface: Surface, options: TabsOptions) -> None:
    if surface.empty:
        return
    theme = surface.theme
    color = options.color if options.color is not None else theme.accent
    total = sum(string_width(t) + 4 for t in options.tabs)
    if options.align == Align.CENTER:
        x = max(0, (surface.width - total) // 2)
    elif options.align == Align.RIGHT:
        x = max(0, surface.width - total)
    else:
        x = 0

    for i, tab in enumerate(options.tabs):
        label = f"  {tab}  "
        if i == options.active:
            if options.variant == TabVariant.UNDERLINE:
                style = Style(fg=color, attrs=Attrs.BOLD | Attrs.UNDERLINE)
            else:
                style = Style(
                    fg=theme.background if theme.dark else theme.surface,
                    bg=color, attrs=Attrs.BOLD,
                )
        else:
            style = Style(fg=theme.muted)
        x += surface.text(
            x, 0, label, TextOptions(fg=style.fg, bg=style.bg, attrs=style.attrs)
        )


@dataclass(frozen=True, slots=True)
class ModalButton:
    label: str = ""
    variant: "ButtonVariant | str" = ButtonVariant.PRIMARY
    focused: bool = False


@dataclass(frozen=True, slots=True)
class ModalOptions:
    title: str = ""
    message: str = ""
    width: int | None = None
    height: int | None = None
    backdrop: bool = True
    """Dim the screen behind the dialog."""
    buttons: Sequence[ModalButton] = ()
    color: Color | None = None
    align: "Align | str" = Align.CENTER


def draw_modal(root: Surface, options: ModalOptions) -> Surface:
    """Centres a dialog over the whole surface and returns its interior, so
    callers can draw custom content instead of ``message`` if they want to."""
    theme = root.theme
    if options.backdrop:
        # Dim rather than blank: the dashboard stays legible behind the dialog.
        root.style_rect(
            0, 0, root.width, root.height,
            Style(fg=theme.foreground.mix(theme.background, 0.72)),
        )

    width = min(options.width if options.width is not None else 48, max(0, root.width - 2))
    message_lines = len(wrap(options.message, max(0, width - 4))) if options.message else 0
    default_height = message_lines + (5 if options.buttons else 4)
    height = min(
        options.height if options.height is not None else default_height,
        max(0, root.height - 2),
    )
    x = max(0, (root.width - width) // 2)
    y = max(0, (root.height - height) // 2)

    surface = root.sub(x, y, width, height)
    inner = surface.box(
        BoxOptions(
            title=options.title, title_align=options.align, border=BorderStyle.ROUNDED,
            border_color=options.color if options.color is not None else theme.border_focused,
            bg=elevate(theme, 0.08),
        )
    )

    if options.message:
        for i, line in enumerate(wrap(options.message, max(0, inner.width - 2))):
            if i + 1 >= inner.height:
                break
            inner.text(
                1, i + 1, fit(line, max(0, inner.width - 2), options.align),
                TextOptions(fg=theme.foreground),
            )

    if options.buttons:
        widths = [string_width(b.label) + 4 for b in options.buttons]
        total = sum(w + 2 for w in widths) - 2
        bx = max(0, (inner.width - total) // 2)
        by = inner.height - 2
        for i, button in enumerate(options.buttons):
            draw_button(
                inner.sub(bx, by, widths[i], 1),
                ButtonOptions(
                    label=button.label, variant=button.variant,
                    focused=button.focused, width=widths[i],
                ),
            )
            bx += widths[i] + 2

    return inner


@dataclass(frozen=True, slots=True)
class PaletteItem:
    label: str = ""
    hint: str = ""


@dataclass(frozen=True, slots=True)
class CommandPaletteOptions:
    query: str = ""
    items: Sequence[PaletteItem] = ()
    selected: int = 0
    width: int | None = None
    height: int | None = None
    placeholder: str = ""


def draw_command_palette(root: Surface, options: CommandPaletteOptions) -> None:
    """Ctrl+K style palette: a query line above a filtered list."""
    theme = root.theme
    width = min(options.width if options.width is not None else 60, max(0, root.width - 2))
    default_height = min(len(options.items) + 4, 14)
    height = min(
        options.height if options.height is not None else default_height,
        max(0, root.height - 2),
    )
    x = max(0, (root.width - width) // 2)
    y = max(1, root.height // 5)

    root.style_rect(
        0, 0, root.width, root.height,
        Style(fg=theme.foreground.mix(theme.background, 0.7)),
    )
    surface = root.sub(x, y, width, height)
    inner = surface.box(
        BoxOptions(
            border=BorderStyle.ROUNDED, border_color=theme.border_focused,
            bg=elevate(theme, 0.1), title="Command Palette",
        )
    )

    inner.text(0, 0, "› ", TextOptions(fg=theme.accent, attrs=Attrs.BOLD))
    query = options.query or options.placeholder or "Type a command…"
    inner.text(
        2, 0, query,
        TextOptions(fg=theme.foreground if options.query else theme.muted),
    )
    inner.hline(0, 1, inner.width, "─", Style(fg=theme.border))

    for i in range(max(0, inner.height - 2)):
        if i >= len(options.items):
            break
        item = options.items[i]
        selected = i == options.selected
        yy = i + 2
        if selected:
            inner.fill_rect(0, yy, inner.width, 1, Style(bg=theme.selection))
        inner.text(
            1, yy, truncate(item.label, max(0, inner.width - 2)),
            TextOptions(
                fg=theme.selection_text if selected else theme.foreground,
                bg=theme.selection if selected else None,
                attrs=Attrs.BOLD if selected else 0,
            ),
        )
        if item.hint:
            hw = string_width(item.hint)
            if hw + 3 < inner.width:
                inner.text(
                    inner.width - hw - 1, yy, item.hint,
                    TextOptions(
                        fg=theme.muted, bg=theme.selection if selected else None
                    ),
                )


@dataclass(frozen=True, slots=True)
class TooltipOptions:
    text: str = ""
    x: int = 0
    y: int = 0
    color: Color | None = None


def draw_tooltip(root: Surface, options: TooltipOptions) -> None:
    theme = root.theme
    width = min(string_width(options.text) + 4, root.width)
    x = max(0, min(options.x, root.width - width))
    y = max(0, min(options.y, root.height - 3))
    surface = root.sub(x, y, width, 3)
    inner = surface.box(
        BoxOptions(
            border=BorderStyle.ROUNDED,
            border_color=options.color if options.color is not None else theme.border_focused,
            bg=elevate(theme, 0.12),
        )
    )
    inner.text(0, 0, truncate(options.text, inner.width), TextOptions(fg=theme.foreground))

File: ports/zig/src/widgets/controls.zig
//! Buttons, inputs, tabs and the overlays that sit above a whole screen.

const std = @import("std");

const buffer_mod = @import("../buffer.zig");
const color_mod = @import("../color.zig");
const surface_mod = @import("../surface.zig");
const theme_mod = @import("../theme.zig");
const unicode = @import("../unicode.zig");

const Align = unicode.Align;
const Attrs = buffer_mod.Attrs;
const BorderStyle = surface_mod.BorderStyle;
const BoxOptions = surface_mod.BoxOptions;
const Color = color_mod.Color;
const Style = buffer_mod.Style;
const Surface = surface_mod.Surface;
const TextOptions = surface_mod.TextOptions;

pub const ButtonVariant = enum { primary, success, warning, danger, ghost };

pub const ButtonOptions = struct {
    label: []const u8 = "",
    focused: bool = false,
    color: ?Color = null,
    variant: ButtonVariant = .primary,
    disabled: bool = false,
    width: ?usize = null,
    alignment: Align = .center,
};

fn variantColor(s: Surface, options: ButtonOptions) Color {
    if (options.color) |c| return c;
    const t = s.theme;
    return switch (options.variant) {
        .success => t.success,
        .warning => t.warning,
        .danger => t.danger,
        .ghost => t.muted,
        .primary => t.primary,
    };
}

/// Returns the width drawn.
pub fn drawButton(s: Surface, options: ButtonOptions) usize {
    if (s.isEmpty()) return 0;
    const theme = s.theme;
    const c = variantColor(s, options);

    var label_buf: [256]u8 = undefined;
    const label = std.fmt.bufPrint(&label_buf, " {s} ", .{options.label}) catch return 0;
    const width = @min(options.width orelse unicode.stringWidth(label), s.width());

    const style: Style = if (options.disabled)
        .{ .fg = theme.muted, .bg = theme_mod.elevate(theme.*, 0.05) }
    else if (options.focused)
        .{
            .fg = if (theme.dark) theme.background else theme.surface,
            .bg = c,
            .attrs = .bold_only,
        }
    else if (options.variant == .ghost)
        .{ .fg = c }
    else
        .{ .fg = c, .bg = theme.surface.mix(c, 0.16), .attrs = .bold_only };

    var cut: [256]u8 = undefined;
    var padded: [256]u8 = undefined;
    const shown = unicode.fit(
        &padded,
        unicode.truncate(&cut, label, width),
        width,
        options.alignment,
    );
    _ = s.text(0, 0, shown, TextOptions.fromStyle(style));
    return width;
}

/// Render as a box, a switch, or a radio dot.
pub const CheckboxVariant = enum { checkbox, toggle, radio };

pub const CheckboxOptions = struct {
    label: []const u8 = "",
    checked: bool = false,
    focused: bool = false,
    color: ?Color = null,
    variant: CheckboxVariant = .checkbox,
};

/// Returns the width drawn.
pub fn drawCheckbox(s: Surface, options: CheckboxOptions) usize {
    if (s.isEmpty()) return 0;
    const theme = s.theme;
    const c = options.color orelse (if (options.checked) theme.success else theme.muted);

    const glyph: []const u8 = switch (options.variant) {
        .toggle => if (options.checked) "[▮ ]" else "[ ▮]",
        .radio => if (options.checked) "(●)" else "( )",
        .checkbox => if (options.checked) "[✓]" else "[ ]",
    };

    const attrs: Attrs = if (options.focused) .bold_only else .none;
    var x = s.text(0, 0, glyph, .{ .fg = c, .attrs = attrs });
    if (options.label.len > 0) {
        var label_buf: [256]u8 = undefined;
        const labelled = std.fmt.bufPrint(&label_buf, " {s}", .{options.label}) catch return x;
        x += s.text(@intCast(x), 0, labelled, .{
            .fg = if (options.focused) theme.foreground else theme.muted,
            .attrs = attrs,
        });
    }
    return x;
}

pub const SelectOptions = struct {
    value: []const u8 = "",
    focused: bool = false,
    open: bool = false,
    options: []const []const u8 = &.{},
    selected_index: usize = 0,
    width: ?usize = null,
    color: ?Color = null,
};

/// A closed dropdown, or an open one with its option list underneath.
pub fn drawSelect(s: Surface, options: SelectOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;
    const width = @min(options.width orelse s.width(), s.width());
    const c = options.color orelse
        (if (options.focused) theme.border_focused else theme.border);
    const field_bg = theme_mod.elevate(theme.*, 0.05);

    var value_cut: [256]u8 = undefined;
    var label_buf: [256]u8 = undefined;
    const value = unicode.truncate(&value_cut, options.value, width -| 4);
    const label = std.fmt.bufPrint(&label_buf, " {s}", .{value}) catch " ";

    var padded: [256]u8 = undefined;
    _ = s.text(0, 0, unicode.fit(&padded, label, width -| 2, .left), .{
        .fg = theme.foreground,
        .bg = field_bg,
        .attrs = if (options.focused) Attrs.bold_only else Attrs.none,
    });
    _ = s.text(
        @as(isize, @intCast(width)) - 2,
        0,
        if (options.open) " ▴" else " ▾",
        .{ .fg = c, .bg = field_bg },
    );

    if (options.open and options.options.len > 0) {
        const height = @min(options.options.len, s.height() -| 1);
        const list_bg = theme_mod.elevate(theme.*, 0.08);
        for (0..height) |i| {
            const selected = i == options.selected_index;
            var item_cut: [256]u8 = undefined;
            var item_buf: [256]u8 = undefined;
            var item_padded: [256]u8 = undefined;
            const item = unicode.truncate(&item_cut, options.options[i], width -| 2);
            const spaced = std.fmt.bufPrint(&item_buf, " {s}", .{item}) catch " ";
            _ = s.text(0, @intCast(i + 1), unicode.fit(&item_padded, spaced, width, .left), .{
                .fg = if (selected) theme.selection_text else theme.foreground,
                .bg = if (selected) theme.selection else list_bg,
            });
        }
    }
}

pub const TextInputOptions = struct {
    value: []const u8 = "",
    placeholder: []const u8 = "",
    focused: bool = false,
    /// Caret index; null means the end of the value.
    cursor: ?usize = null,
    width: ?usize = null,
    label: []const u8 = "",
    password: bool = false,
    color: ?Color = null,
};

pub fn drawTextInput(s: Surface, options: TextInputOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;
    const width = @min(options.width orelse s.width(), s.width());
    const label_width: usize = if (options.label.len > 0)
        unicode.stringWidth(options.label) + 1
    else
        0;
    if (options.label.len > 0) {
        _ = s.text(0, 0, options.label, .{ .fg = theme.muted });
    }
    const field_width = width -| label_width;
    const bg = theme_mod.elevate(theme.*, if (options.focused) 0.1 else 0.05);
    s.fillRect(@intCast(label_width), 0, field_width, 1, .{ .bg = bg }, 32);

    // The reference counts UTF-16 units for the password mask and the default
    // caret; counting codepoints is the same for every value a person types and
    // is what a Zig caller would expect.
    var mask_buf: [256]u8 = undefined;
    var shown = options.value;
    if (options.password) {
        var n: usize = 0;
        var it = std.unicode.Utf8Iterator{ .bytes = options.value, .i = 0 };
        while (it.nextCodepoint()) |_| : (n += 1) {
            if ((n + 1) * 3 > mask_buf.len) break;
            @memcpy(mask_buf[n * 3 ..][0..3], "•");
        }
        shown = mask_buf[0 .. n * 3];
    }
    const empty = shown.len == 0;
    const text = if (empty) options.placeholder else shown;

    var cut: [256]u8 = undefined;
    _ = s.text(
        @as(isize, @intCast(label_width)) + 1,
        0,
        unicode.truncate(&cut, text, field_width -| 2),
        .{ .fg = if (empty) theme.muted else theme.foreground, .bg = bg },
    );

    if (options.focused) {
        const caret = options.cursor orelse unicode.stringWidth(shown);
        const cursor_x = @min(label_width + 1 + caret, label_width + (field_width -| 1));
        s.styleRect(@intCast(cursor_x), 0, 1, 1, .{
            .fg = theme.background,
            .bg = options.color orelse theme.cursor,
        });
    }
}

/// Underline the active tab instead of filling it.
pub const TabVariant = enum { filled, underline };

pub const TabsOptions = struct {
    tabs: []const []const u8 = &.{},
    active: usize = 0,
    color: ?Color = null,
    alignment: Align = .left,
    variant: TabVariant = .filled,
};

pub fn drawTabs(s: Surface, options: TabsOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;
    const c = options.color orelse theme.accent;

    var total: usize = 0;
    for (options.tabs) |t| total += unicode.stringWidth(t) + 4;

    const width: isize = @intCast(s.width());
    var x: isize = switch (options.alignment) {
        .center => @max(0, @divFloor(width - @as(isize, @intCast(total)), 2)),
        .right => @max(0, width - @as(isize, @intCast(total))),
        .left => 0,
    };

    for (options.tabs, 0..) |tab, i| {
        var label_buf: [256]u8 = undefined;
        const label = std.fmt.bufPrint(&label_buf, "  {s}  ", .{tab}) catch continue;
        const style: Style = if (i == options.active) switch (options.variant) {
            .underline => .{ .fg = c, .attrs = .{ .bold = true, .underline = true } },
            .filled => .{
                .fg = if (theme.dark) theme.background else theme.surface,
                .bg = c,
                .attrs = .bold_only,
            },
        } else .{ .fg = theme.muted };
        x += @intCast(s.text(x, 0, label, TextOptions.fromStyle(style)));
    }
}

pub const ModalButton = struct {
    label: []const u8 = "",
    variant: ButtonVariant = .primary,
    focused: bool = false,
};

pub const ModalOptions = struct {
    title: []const u8 = "",
    message: []const u8 = "",
    width: ?usize = null,
    height: ?usize = null,
    /// Dim the screen behind the dialog.
    backdrop: bool = true,
    buttons: []const ModalButton = &.{},
    color: ?Color = null,
    alignment: Align = .center,
};

/// Centres a dialog over the whole surface and returns its interior, so callers
/// can draw custom content instead of `message` if they want to.
pub fn drawModal(allocator: std.mem.Allocator, root: Surface, options: ModalOptions) !Surface {
    const theme = root.theme;
    if (options.backdrop) {
        // Dim rather than blank: the dashboard stays legible behind the dialog.
        root.styleRect(0, 0, root.width(), root.height(), .{
            .fg = theme.foreground.mix(theme.background, 0.72),
        });
    }

    const width = @min(options.width orelse 48, root.width() -| 2);

    var message_lines: usize = 0;
    if (options.message.len > 0) message_lines = unicode.wrapCount(options.message, width -| 4);
    const default_height = message_lines + @as(usize, if (options.buttons.len > 0) 5 else 4);
    const height = @min(options.height orelse default_height, root.height() -| 2);

    const x = @max(0, @divFloor(@as(isize, @intCast(root.width())) - @as(isize, @intCast(width)), 2));
    const y = @max(0, @divFloor(@as(isize, @intCast(root.height())) - @as(isize, @intCast(height)), 2));

    const s = root.sub(x, y, width, height);
    const inner = s.box(.{
        .title = options.title,
        .title_align = options.alignment,
        .border = .rounded,
        .border_color = options.color orelse theme.border_focused,
        .bg = theme_mod.elevate(theme.*, 0.08),
    });

    if (options.message.len > 0) {
        const lines = try unicode.wrap(allocator, options.message, inner.width() -| 2);
        defer unicode.freeWrapped(allocator, lines);
        for (lines, 0..) |line, i| {
            if (i + 1 >= inner.height()) break;
            var padded: [512]u8 = undefined;
            _ = inner.text(
                1,
                @intCast(i + 1),
                unicode.fit(&padded, line, inner.width() -| 2, options.alignment),
                .{ .fg = theme.foreground },
            );
        }
    }

    if (options.buttons.len > 0) {
        var total: isize = -2;
        for (options.buttons) |b| total += @intCast(unicode.stringWidth(b.label) + 4 + 2);
        var bx: isize = @max(0, @divFloor(@as(isize, @intCast(inner.width())) - total, 2));
        const by: isize = @as(isize, @intCast(inner.height())) - 2;
        for (options.buttons) |button| {
            const bw = unicode.stringWidth(button.label) + 4;
            _ = drawButton(inner.sub(bx, by, bw, 1), .{
                .label = button.label,
                .variant = button.variant,
                .focused = button.focused,
                .width = bw,
            });
            bx += @as(isize, @intCast(bw)) + 2;
        }
    }

    return inner;
}

pub const PaletteItem = struct {
    label: []const u8 = "",
    hint: []const u8 = "",
};

pub const CommandPaletteOptions = struct {
    query: []const u8 = "",
    items: []const PaletteItem = &.{},
    selected: usize = 0,
    width: ?usize = null,
    height: ?usize = null,
    placeholder: []const u8 = "",
};

/// Ctrl+K style palette: a query line above a filtered list.
pub fn drawCommandPalette(root: Surface, options: CommandPaletteOptions) void {
    const theme = root.theme;
    const width = @min(options.width orelse 60, root.width() -| 2);
    const default_height = @min(options.items.len + 4, 14);
    const height = @min(options.height orelse default_height, root.height() -| 2);

    const x = @max(0, @divFloor(@as(isize, @intCast(root.width())) - @as(isize, @intCast(width)), 2));
    const y = @max(1, @as(isize, @intCast(root.height() / 5)));

    root.styleRect(0, 0, root.width(), root.height(), .{
        .fg = theme.foreground.mix(theme.background, 0.7),
    });

    const s = root.sub(x, y, width, height);
    const inner = s.box(.{
        .border = .rounded,
        .border_color = theme.border_focused,
        .bg = theme_mod.elevate(theme.*, 0.1),
        .title = "Command Palette",
    });

    _ = inner.text(0, 0, "› ", .{ .fg = theme.accent, .attrs = .bold_only });
    const query_empty = options.query.len == 0;
    const query = if (!query_empty)
        options.query
    else if (options.placeholder.len > 0)
        options.placeholder
    else
        "Type a command…";
    _ = inner.text(2, 0, query, .{
        .fg = if (query_empty) theme.muted else theme.foreground,
    });
    inner.hline(0, 1, inner.width(), '─', .{ .fg = theme.border });

    const list_height = inner.height() -| 2;
    for (0..list_height) |i| {
        if (i >= options.items.len) break;
        const item = options.items[i];
        const selected = i == options.selected;
        const yy: isize = @intCast(i + 2);
        if (selected) inner.fillRect(0, yy, inner.width(), 1, .{ .bg = theme.selection }, 32);

        var cut: [256]u8 = undefined;
        _ = inner.text(1, yy, unicode.truncate(&cut, item.label, inner.width() -| 2), .{
            .fg = if (selected) theme.selection_text else theme.foreground,
            .bg = if (selected) theme.selection else null,
            .attrs = if (selected) Attrs.bold_only else Attrs.none,
        });

        if (item.hint.len > 0) {
            const hw = unicode.stringWidth(item.hint);
            if (hw + 3 < inner.width()) {
                _ = inner.text(@intCast(inner.width() - hw - 1), yy, item.hint, .{
                    .fg = theme.muted,
                    .bg = if (selected) theme.selection else null,
                });
            }
        }
    }
}

pub const TooltipOptions = struct {
    text: []const u8 = "",
    x: isize = 0,
    y: isize = 0,
    color: ?Color = null,
};

pub fn drawTooltip(root: Surface, options: TooltipOptions) void {
    const theme = root.theme;
    const width = @min(unicode.stringWidth(options.text) + 4, root.width());
    const x = @max(0, @min(options.x, @as(isize, @intCast(root.width())) - @as(isize, @intCast(width))));
    const y = @max(0, @min(options.y, @as(isize, @intCast(root.height())) - 3));

    const s = root.sub(x, y, width, 3);
    const inner = s.box(.{
        .border = .rounded,
        .border_color = options.color orelse theme.border_focused,
        .bg = theme_mod.elevate(theme.*, 0.12),
    });
    var cut: [256]u8 = undefined;
    _ = inner.text(0, 0, unicode.truncate(&cut, options.text, inner.width()), .{
        .fg = theme.foreground,
    });
}
